src/libstd/sys/unix/fs.rs

```diff
@@ -463,18 +468,24 @@ pub fn rmdir(p: &Path) -> io::Result<()> {
 pub fn readlink(p: &Path) -> io::Result<PathBuf> {
     let c_path = try!(cstr(p));
     let p = c_path.as_ptr();
-    let mut len = unsafe { libc::pathconf(p as *mut _, libc::_PC_NAME_MAX) };
-    if len < 0 {
-        len = 1024; // FIXME: read PATH_MAX from C ffi?
-    }
-    let mut buf: Vec<u8> = Vec::with_capacity(len as usize);
-    unsafe {
-        let n = try!(cvt({
-            libc::readlink(p, buf.as_ptr() as *mut c_char, len as size_t)
-        }));
-        buf.set_len(n as usize);
+
+    let mut buf = Vec::with_capacity(256);
+
+    loop {
+        let buf_read = try!(cvt(unsafe { libc::readlink(
+                                                    p,
+                                                    buf.as_mut_ptr() as *mut libc::c_char,
+                                                    buf.capacity() as libc::size_t) })) as usize;
+        if buf_read != buf.capacity() {
+            unsafe { buf.set_len(buf_read); }
+            buf.shrink_to_fit();
+
+            return Ok(PathBuf::from(OsString::from_vec(buf)));
+        }
+
+        let new_cap = buf.capacity() * 2;
+        buf.reserve(new_cap);
```

Comment:
The length of `buf` remains 0 until the end, so I think `buf.reserve()` will work as expected in this case. The problem is, as you said, `Vec::reserve` has the internal doubling logic, so the capacity will grow as in `N, 4N, 16N, 64N, ...`.

The recently added [getcwd](https://github.com/rust-lang/rust/blob/master/src/libstd/sys/unix/os.rs#L101-L114) has a similar logic but it uses an external variable to track the desired capacity, so the capacity being four times each iteration will not happen. However, every one of two `Vec::reserve` calls will be no-op because the previously increased capacity will be sufficient to meet the new requirement. This also means every one of two system calls is wasted.

Correct me if I'm wrong, but the behavior of `read_to_end_uninitialized` seems to be a bit misleading too. What it will do is `N, 2(N+1), 2(2(N+1))+1, ...`, because the internal doubling is applied after the capacity is increased by one. Is this an expected behavior? I don't know much about the allocator, but I guess the capacity remaining `2^N` seems to be ideal. The behavior does not follow this.

There seems to be a more appropriate function for this purpose: `Vec::reserve_exact`. It has the almost identical behavior except it does not double the capacity internally, so it would be useful if the user is assured to provide an appropriate new capacity and also wants to prevent excessive memory allocation. What do you think of this?